service/instance.py

```diff
@@ -1319,57 +1321,111 @@ def user_delete_security_group(core_identity):
+            "port_range_min": 22,
+            "port_range_max": 22,
+            "protocol": "tcp",
+            "remote_ip_prefix": "0.0.0.0/0",
+        }
+    ]
+    return extended_default_rules
+
+
+def neutron_set_security_group_rules(security_group_name, security_group_rules_dict, user_neutron):
+    security_group = find_security_group(security_group_name, user_neutron)
+    security_group_id = security_group[u'id']
+    for sg_rule in security_group_rules_dict:
         try:
-            # attempt to find
-            rule_found = any(
-                sg_rule for sg_rule in security_group.rules
-                if sg_rule.ip_protocol == ip_protocol.lower() and
-                sg_rule.from_port == from_port and
-                sg_rule.to_port == to_port and
-                (not cidr or sg_rule.ip_range == cidr))
-            if rule_found:
-                continue
-            # Attempt to create
-            lc_driver.ex_create_security_group_rule(security_group, ip_protocol, from_port, to_port, cidr)
-        except BaseHTTPError as exc:
-            if "Security group rule already exists" in exc.message:
-                continue
-            raise
-    return security_group
+            rule_body = {"security_group_rule": {
+                "direction": sg_rule['direction'],
+                "port_range_min": sg_rule.get('port_range_min', None),
+                "ethertype": sg_rule.get("ethertype", "IPv4"),
+                "port_range_max": sg_rule.get('port_range_max', None),
+                "protocol": sg_rule.get("protocol", None),
+                "remote_group_id": security_group_id,
+                "security_group_id": security_group_id}
+            }
+            if 'remote_ip_prefix' in sg_rule:
+                rule_body['security_group_rule']['remote_ip_prefix'] = sg_rule['remote_ip_prefix']
+                rule_body['security_group_rule'].pop("remote_group_id")
+            user_neutron.create_security_group_rule(body=rule_body)
+        except Conflict:
+            pass
+    return True
+
+
+def find_security_group(security_group_name, user_neutron):
+    security_groups = user_neutron.list_security_groups()[u'security_groups']
+    security_group = ''
+    for sg in security_groups:
+        if sg[u'name'] == security_group_name:
+            security_group = sg
+    if security_group != '':
+        return security_group
+    else:
+        raise Exception('Could not find any existing security group')
 
 
-def get_or_create_security_group(lc_driver, security_group_name):
-    sgroup_list = lc_driver.ex_list_security_groups()
-    security_group = [sgroup for sgroup in sgroup_list if sgroup.name == security_group_name]
+def get_or_create_security_group(security_group_name, user_neutron):
+    security_group_list = user_neutron.list_security_groups()[u'security_groups']
+    security_group = [sgroup for sgroup in security_group_list if sgroup[u'name'] == security_group_name]
```

Comment:
This function is a upstream function (written by atmosphere folks) and it tends to `get` a security group first. If the security group doesn't exist, it will create one. However, the find_security_group is trying to `get` a security group only. I got your point but do you think I could combine these two functions?